From 10991ec250d42392b407a5a36ee8649f6932ba97 Mon Sep 17 00:00:00 2001
From: Brendan Dahl <bdahl@mozilla.com>
Date: Wed, 31 Oct 2018 15:59:40 -0700
Subject: [PATCH] Disable xml content sink interrupt flush.

---
 dom/xml/nsXMLContentSink.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/dom/xml/nsXMLContentSink.cpp b/dom/xml/nsXMLContentSink.cpp
index e651e95387b8..bc10abfec03e 100644
--- a/dom/xml/nsXMLContentSink.cpp
+++ b/dom/xml/nsXMLContentSink.cpp
@@ -468,13 +468,15 @@ nsXMLContentSink::StyleSheetLoaded(StyleSheet* aSheet,
 NS_IMETHODIMP
 nsXMLContentSink::WillInterrupt(void)
 {
-  return WillInterruptImpl();
+  // return WillInterruptImpl();
+  return NS_OK;
 }
 
 NS_IMETHODIMP
 nsXMLContentSink::WillResume(void)
 {
-  return WillResumeImpl();
+  // return WillResumeImpl();
+  return NS_OK;
 }
 
 NS_IMETHODIMP
-- 
2.17.1

