diff --git a/toolkit/content/widgets/button.js b/toolkit/content/widgets/button.js
--- a/toolkit/content/widgets/button.js
+++ b/toolkit/content/widgets/button.js
@@ -104,11 +104,11 @@
             return;
           }
         }
 
         // Test dialog buttons
-        let buttonBox = window.top.document.documentElement.buttonBox;
+        let buttonBox = window.top.document.querySelector("dialog")?.buttonBox;
         if (buttonBox) {
           this.fireAccessKeyButton(buttonBox, charPressedLower);
         }
       });
     }

