diff --git a/testing/mozharness/mozharness/mozilla/testing/raptor.py b/testing/mozharness/mozharness/mozilla/testing/raptor.py
--- a/testing/mozharness/mozharness/mozilla/testing/raptor.py
+++ b/testing/mozharness/mozharness/mozilla/testing/raptor.py
@@ -612,10 +612,45 @@
                         "pageload metrics."
                         "Options: geomean, mean."
                     ),
                 },
             ],
+            [
+                ["--benchmark-repository"],
+                {
+                    "dest": "benchmark_repository",
+                    "type": "str",
+                    "default": None,
+                    "help": (
+                        "Repository that should be used for a particular benchmark test. "
+                        "e.g. https://github.com/mozilla-mobile/firefox-android"
+                    ),
+                },
+            ],
+            [
+                ["--benchmark-revision"],
+                {
+                    "dest": "benchmark_revision",
+                    "type": "str",
+                    "default": None,
+                    "help": (
+                        "Repository revision that should be used for a particular "
+                        "benchmark test."
+                    ),
+                },
+            ],
+            [
+                ["--benchmark-branch"],
+                {
+                    "dest": "benchmark_branch",
+                    "type": "str",
+                    "default": None,
+                    "help": (
+                        "Repository branch that should be used for a particular benchmark test."
+                    ),
+                },
+            ],
         ]
         + testing_config_options
         + copy.deepcopy(code_coverage_config_options)
         + browsertime_options
     )
@@ -971,10 +1006,16 @@
             kw_options["device-name"] = self.config["device_name"]
         if self.config.get("activity") is not None:
             kw_options["activity"] = self.config["activity"]
         if self.config.get("conditioned_profile") is not None:
             kw_options["conditioned-profile"] = self.config["conditioned_profile"]
+        if self.config.get("benchmark_repository"):
+            kw_options["benchmark_repository"] = self.config["benchmark_repository"]
+        if self.config.get("benchmark_revision"):
+            kw_options["benchmark_revision"] = self.config["benchmark_revision"]
+        if self.config.get("benchmark_repository"):
+            kw_options["benchmark_branch"] = self.config["benchmark_branch"]
 
         kw_options.update(kw)
         if self.host:
             kw_options["host"] = self.host
         # Configure profiling options
diff --git a/testing/raptor/raptor/benchmark.py b/testing/raptor/raptor/benchmark.py
--- a/testing/raptor/raptor/benchmark.py
+++ b/testing/raptor/raptor/benchmark.py
@@ -22,10 +22,20 @@
 
     def __init__(self, config, test):
         self.config = config
         self.test = test
 
+        # Note that we can only change the repository, revision, and branch through here.
+        # The path to the test should remain constant. If it needs to be changed, make a
+        # patch that changes it for the new test.
+        if self.config.get("benchmark_repository", None):
+            self.test["repository"] = self.config["benchmark_repository"]
+            self.test["repository_revision"] = self.config["benchmark_revision"]
+
+            if self.config.get("benchmark_branch", None):
+                self.test["branch"] = self.config["benchmark_branch"]
+
         self.setup_benchmarks(
             os.getenv("MOZ_DEVELOPER_REPO_DIR"),
             os.getenv("MOZ_MOZBUILD_DIR"),
             run_local=self.config.get("run_local", False),
         )
diff --git a/testing/raptor/raptor/cmdline.py b/testing/raptor/raptor/cmdline.py
--- a/testing/raptor/raptor/cmdline.py
+++ b/testing/raptor/raptor/cmdline.py
@@ -511,10 +511,32 @@
         default=[],
         metavar="OPTION",
         help="Alternative methods for summarizing technical and visual pageload metrics. "
         "Options: geomean, mean.",
     )
+    add_arg(
+        "--benchmark-repository",
+        dest="benchmark_repository",
+        default=None,
+        type=str,
+        help="Repository that should be used for a particular benchmark test. "
+        "e.g. https://github.com/mozilla-mobile/firefox-android",
+    )
+    add_arg(
+        "--benchmark-revision",
+        dest="benchmark_revision",
+        default=None,
+        type=str,
+        help="Repository revision that should be used for a particular benchmark test.",
+    )
+    add_arg(
+        "--benchmark-branch",
+        dest="benchmark_branch",
+        default=None,
+        type=str,
+        help="Repository branch that should be used for a particular benchmark test.",
+    )
 
     add_logging_group(parser)
     return parser
 
 
@@ -616,10 +638,16 @@
                 args.intent = APPS[args.app]["default_intent"]
             else:
                 # otherwise fail out
                 parser.error("--intent command-line argument is required!")
 
+    if args.benchmark_repository:
+        if not args.benchmark_revision:
+            parser.error(
+                "When a benchmark repository is provided, a revision is also required."
+            )
+
 
 def parse_args(argv=None):
     parser = create_parser()
     args = parser.parse_args(argv)
     if args.host == "HOST_IP":
diff --git a/testing/raptor/raptor/perftest.py b/testing/raptor/raptor/perftest.py
--- a/testing/raptor/raptor/perftest.py
+++ b/testing/raptor/raptor/perftest.py
@@ -103,10 +103,13 @@
         project="mozilla-central",
         verbose=False,
         python=None,
         fission=True,
         extra_summary_methods=[],
+        benchmark_repository=None,
+        benchmark_revision=None,
+        benchmark_branch=None,
         **kwargs
     ):
         self._remote_test_root = None
         self._dirs_to_remove = []
         self.verbose = verbose
@@ -149,10 +152,13 @@
             "extra_prefs": extra_prefs,
             "environment": environment,
             "project": project,
             "verbose": verbose,
             "extra_summary_methods": extra_summary_methods,
+            "benchmark_repository": benchmark_repository,
+            "benchmark_revision": benchmark_revision,
+            "benchmark_branch": benchmark_branch,
         }
 
         self.firefox_android_apps = FIREFOX_ANDROID_APPS
 
         # We are deactivating the conditioned profiles for:
diff --git a/testing/raptor/raptor/raptor.py b/testing/raptor/raptor/raptor.py
--- a/testing/raptor/raptor/raptor.py
+++ b/testing/raptor/raptor/raptor.py
@@ -136,10 +136,13 @@
             chimera=args.chimera,
             project=args.project,
             verbose=args.verbose,
             fission=args.fission,
             extra_summary_methods=args.extra_summary_methods,
+            benchmark_repository=args.benchmark_repository,
+            benchmark_revision=args.benchmark_revision,
+            benchmark_branch=args.benchmark_branch,
         )
     except Exception:
         traceback.print_exc()
         LOG.critical(
             "TEST-UNEXPECTED-FAIL: could not initialize the raptor test runner"

