diff --git a/browser/app/profile/firefox.js b/browser/app/profile/firefox.js
--- a/browser/app/profile/firefox.js
+++ b/browser/app/profile/firefox.js
@@ -1406,10 +1406,13 @@
 
 // URL for "Learn More" for Crash Reporter
 pref("toolkit.crashreporter.infoURL",
      "https://www.mozilla.org/legal/privacy/firefox.html#crash-reporter");
 
+// URL for "Learn More" for Private Attribution
+pref("dom.private-attribution.infoURL", "https://support.mozilla.org/1/firefox/%VERSION%/%OS%/%LOCALE%/privacy-preserving-attribution");
+
 // base URL for web-based support pages
 pref("app.support.baseURL", "https://support.mozilla.org/1/firefox/%VERSION%/%OS%/%LOCALE%/");
 
 // base url for web-based feedback pages
 pref("app.feedback.baseURL", "https://ideas.mozilla.org/");
diff --git a/browser/components/preferences/privacy.inc.xhtml b/browser/components/preferences/privacy.inc.xhtml
--- a/browser/components/preferences/privacy.inc.xhtml
+++ b/browser/components/preferences/privacy.inc.xhtml
@@ -1029,10 +1029,29 @@
                value="false"/>
     </radiogroup>
   </vbox>
 </groupbox>
 
+#ifdef MOZ_DATA_REPORTING
+<!-- Website Advertising Preferences -->
+<hbox id="websiteAdvertisingCategory"
+      class="subcategory"
+      hidden="true"
+      data-category="panePrivacy">
+  <html:h1 data-l10n-id="website-advertising-header"/>
+</hbox>
+
+<groupbox id="websiteAdvertisingGroup" data-category="panePrivacy" hidden="true">
+  <checkbox id="privateAttribution"
+            preference="dom.private-attribution.submission.enabled"
+            data-l10n-id="website-advertising-private-attribution"/>
+  <description class="indent tip-caption" data-l10n-id="website-advertising-private-attribution-description">
+    <html:a id="privateAttributionLearnMore" is="moz-support-link" class="learnMore" data-l10n-name="learn-more"/>
+  </description>
+</groupbox>
+#endif
+
 <hbox id="securityCategory"
       class="subcategory"
       hidden="true"
       data-category="panePrivacy">
   <html:h1 data-l10n-id="security-header"/>
diff --git a/browser/components/preferences/privacy.js b/browser/components/preferences/privacy.js
--- a/browser/components/preferences/privacy.js
+++ b/browser/components/preferences/privacy.js
@@ -276,10 +276,11 @@
   Preferences.addAll([
     // Preference instances for prefs that we need to monitor while the page is open.
     { id: PREF_OPT_OUT_STUDIES_ENABLED, type: "bool" },
     { id: PREF_ADDON_RECOMMENDATIONS_ENABLED, type: "bool" },
     { id: PREF_UPLOAD_ENABLED, type: "bool" },
+    { id: "dom.private-attribution.submission.enabled", type: "bool" },
   ]);
 }
 // Privacy segmentation section
 Preferences.add({
   id: "browser.dataFeatureRecommendations.enabled",
@@ -1214,10 +1215,11 @@
 
     if (AppConstants.MOZ_DATA_REPORTING) {
       if (AppConstants.MOZ_CRASHREPORTER) {
         this.initSubmitCrashes();
       }
+      this.initPrivateAttribution();
       this.initSubmitHealthReport();
       setEventListener(
         "submitHealthReportBox",
         "command",
         gPrivacyPane.updateSubmitHealthReport
@@ -3376,10 +3378,17 @@
       "toolkit.crashreporter.infoURL",
       "crashReporterLearnMore"
     );
   },
 
+  initPrivateAttribution() {
+    this._setupLearnMoreLink(
+      "dom.private-attribution.infoURL",
+      "privateAttributionLearnMore"
+    );
+  },
+
   initPrivacySegmentation() {
     // Section visibility
     let section = document.getElementById("privacySegmentationSection");
     let updatePrivacySegmentationSectionVisibilityState = () => {
       section.hidden = !Services.prefs.getBoolPref(
diff --git a/browser/locales/en-US/browser/preferences/preferences.ftl b/browser/locales/en-US/browser/preferences/preferences.ftl
--- a/browser/locales/en-US/browser/preferences/preferences.ftl
+++ b/browser/locales/en-US/browser/preferences/preferences.ftl
@@ -1473,10 +1473,20 @@
     .label = Use { -brand-product-name } recommendations
 
 privacy-segmentation-radio-on =
     .label = Show detailed information
 
+## Privacy Section - Website Advertising Preferences
+
+website-advertising-header = Website Advertising Preferences
+
+website-advertising-private-attribution =
+    .label = Allow websites to perform privacy-preserving ad measurement
+    .accesskey = a
+
+website-advertising-private-attribution-description = This helps sites understand how their ads perform without collecting data about you. <a data-l10n-name="learn-more">Learn More</a>
+
 ## Privacy Section - Security
 ##
 ## It is important that wording follows the guidelines outlined on this page:
 ## https://developers.google.com/safe-browsing/developers_guide_v2#AcceptableUsage
 
diff --git a/dom/privateattribution/PrivateAttribution.cpp b/dom/privateattribution/PrivateAttribution.cpp
--- a/dom/privateattribution/PrivateAttribution.cpp
+++ b/dom/privateattribution/PrivateAttribution.cpp
@@ -50,10 +50,14 @@
   return true;
 }
 
 void PrivateAttribution::SaveImpression(
     const PrivateAttributionImpressionOptions& aOptions, ErrorResult& aRv) {
+  if (!StaticPrefs::dom_private_attribution_submission_enabled()) {
+    return;
+  }
+
   nsAutoCString source;
   if (!GetSourceHost(source, aRv)) {
     return;
   }
 
@@ -80,10 +84,14 @@
                                 aOptions.mAd, aOptions.mTarget);
 }
 
 void PrivateAttribution::MeasureConversion(
     const PrivateAttributionConversionOptions& aOptions, ErrorResult& aRv) {
+  if (!StaticPrefs::dom_private_attribution_submission_enabled()) {
+    return;
+  }
+
   nsAutoCString source;
   if (!GetSourceHost(source, aRv)) {
     return;
   }
   for (const nsACString& host : aOptions.mSources) {
diff --git a/modules/libpref/init/StaticPrefList.yaml b/modules/libpref/init/StaticPrefList.yaml
--- a/modules/libpref/init/StaticPrefList.yaml
+++ b/modules/libpref/init/StaticPrefList.yaml
@@ -3221,10 +3221,17 @@
 - name: dom.origin-trials.private-attribution.state
   type: RelaxedAtomicInt32
   value: 0
   mirror: always
 
+# User pref to control whether Private Attribution
+# information should be collected / submitted.
+- name: dom.private-attribution.submission.enabled
+  type: bool
+  value: true
+  mirror: always
+
 # Is support for Window.paintWorklet enabled?
 - name: dom.paintWorklet.enabled
   type: bool
   value: false
   mirror: always

