diff --git a/browser/app/profile/firefox.js b/browser/app/profile/firefox.js
--- a/browser/app/profile/firefox.js
+++ b/browser/app/profile/firefox.js
@@ -1399,10 +1399,13 @@
 
 // URL for "Learn More" for Crash Reporter
 pref("toolkit.crashreporter.infoURL",
      "https://www.mozilla.org/legal/privacy/firefox.html#crash-reporter");
 
+// URL for "Learn More" for Private Attribution
+pref("dom.private-attribution.infoURL", "https://support.mozilla.org/1/firefox/%VERSION%/%OS%/%LOCALE%/privacy-preserving-attribution");
+
 // base URL for web-based support pages
 pref("app.support.baseURL", "https://support.mozilla.org/1/firefox/%VERSION%/%OS%/%LOCALE%/");
 
 // base url for web-based feedback pages
 pref("app.feedback.baseURL", "https://ideas.mozilla.org/");
diff --git a/browser/components/preferences/privacy.inc.xhtml b/browser/components/preferences/privacy.inc.xhtml
--- a/browser/components/preferences/privacy.inc.xhtml
+++ b/browser/components/preferences/privacy.inc.xhtml
@@ -997,10 +997,21 @@
                 data-l10n-id="collection-backlogged-crash-reports"
                 data-l10n-attrs="accesskey"/>
       <html:a id="crashReporterLearnMore" is="moz-support-link" class="learnMore"/>
     </hbox>
 #endif
+    <hbox align="center" class="checkbox-row">
+      <html:input type="checkbox"
+                id="privateAttribution"
+                preference="dom.private-attribution.submission.enabled"/>
+      <html:label is="moz-label"
+                for="privateAttribution"
+                id="privateAttributionLabel"
+                data-l10n-id="collection-private-attribution"
+                data-l10n-attrs="accesskey"/>
+      <html:a id="privateAttributionLearnMore" is="moz-support-link" class="learnMore"/>
+    </hbox>
   </vbox>
 #endif
   <vbox class="firefoxSuggestOptionBox" id="firefoxSuggestPrivacyContainer" hidden="true">
     <html:moz-toggle id="firefoxSuggestDataCollectionPrivacyToggle"
                      preference="browser.urlbar.quicksuggest.dataCollection.enabled"
diff --git a/browser/components/preferences/privacy.js b/browser/components/preferences/privacy.js
--- a/browser/components/preferences/privacy.js
+++ b/browser/components/preferences/privacy.js
@@ -276,10 +276,11 @@
   Preferences.addAll([
     // Preference instances for prefs that we need to monitor while the page is open.
     { id: PREF_OPT_OUT_STUDIES_ENABLED, type: "bool" },
     { id: PREF_ADDON_RECOMMENDATIONS_ENABLED, type: "bool" },
     { id: PREF_UPLOAD_ENABLED, type: "bool" },
+    { id: "dom.private-attribution.submission.enabled", type: "bool" },
   ]);
 }
 // Privacy segmentation section
 Preferences.add({
   id: "browser.dataFeatureRecommendations.enabled",
@@ -1198,10 +1199,11 @@
 
     if (AppConstants.MOZ_DATA_REPORTING) {
       if (AppConstants.MOZ_CRASHREPORTER) {
         this.initSubmitCrashes();
       }
+      this.initPrivateAttribution();
       this.initSubmitHealthReport();
       setEventListener(
         "submitHealthReportBox",
         "command",
         gPrivacyPane.updateSubmitHealthReport
@@ -3360,10 +3362,17 @@
       "toolkit.crashreporter.infoURL",
       "crashReporterLearnMore"
     );
   },
 
+  initPrivateAttribution() {
+    this._setupLearnMoreLink(
+      "dom.private-attribution.infoURL",
+      "privateAttributionLearnMore"
+    );
+  },
+
   initPrivacySegmentation() {
     // Section visibility
     let section = document.getElementById("privacySegmentationSection");
     let updatePrivacySegmentationSectionVisibilityState = () => {
       section.hidden = !Services.prefs.getBoolPref(
diff --git a/browser/locales/en-US/browser/preferences/preferences.ftl b/browser/locales/en-US/browser/preferences/preferences.ftl
--- a/browser/locales/en-US/browser/preferences/preferences.ftl
+++ b/browser/locales/en-US/browser/preferences/preferences.ftl
@@ -1463,10 +1463,13 @@
 collection-health-report-disabled = Data reporting is disabled for this build configuration
 
 collection-backlogged-crash-reports = Allow { -brand-short-name } to send backlogged crash reports on your behalf
     .accesskey = c
 
+collection-private-attribution = Allow websites to perform privacy-preserving ad measurement
+    .accesskey = a
+
 privacy-segmentation-section-header = New features that enhance your browsing
 
 privacy-segmentation-section-description = When we offer features that use your data to give you a more personal experience:
 
 privacy-segmentation-radio-off =
diff --git a/dom/privateattribution/PrivateAttribution.cpp b/dom/privateattribution/PrivateAttribution.cpp
--- a/dom/privateattribution/PrivateAttribution.cpp
+++ b/dom/privateattribution/PrivateAttribution.cpp
@@ -50,10 +50,14 @@
   return true;
 }
 
 void PrivateAttribution::SaveImpression(
     const PrivateAttributionImpressionOptions& aOptions, ErrorResult& aRv) {
+  if (!StaticPrefs::dom_private_attribution_submission_enabled()) {
+    return;
+  }
+
   nsAutoCString source;
   if (!GetSourceHost(source, aRv)) {
     return;
   }
 
@@ -80,10 +84,14 @@
                                 aOptions.mAd, aOptions.mTarget);
 }
 
 void PrivateAttribution::MeasureConversion(
     const PrivateAttributionConversionOptions& aOptions, ErrorResult& aRv) {
+  if (!StaticPrefs::dom_private_attribution_submission_enabled()) {
+    return;
+  }
+
   nsAutoCString source;
   if (!GetSourceHost(source, aRv)) {
     return;
   }
   for (const nsACString& host : aOptions.mSources) {
diff --git a/modules/libpref/init/StaticPrefList.yaml b/modules/libpref/init/StaticPrefList.yaml
--- a/modules/libpref/init/StaticPrefList.yaml
+++ b/modules/libpref/init/StaticPrefList.yaml
@@ -3242,10 +3242,17 @@
 - name: dom.origin-trials.private-attribution.state
   type: RelaxedAtomicInt32
   value: 0
   mirror: always
 
+# User pref to control whether Private Attribution
+# information should be collected / submitted.
+- name: dom.private-attribution.submission.enabled
+  type: bool
+  value: true
+  mirror: always
+
 # Is support for Window.paintWorklet enabled?
 - name: dom.paintWorklet.enabled
   type: bool
   value: false
   mirror: always

