# HG changeset patch
# Parent  aae5ea602f13ee4883811646a88601edbde92b83

diff --git a/browser/components/migration/content/migration.xul b/browser/components/migration/content/migration.xul
--- a/browser/components/migration/content/migration.xul
+++ b/browser/components/migration/content/migration.xul
@@ -11,21 +11,21 @@
         xmlns="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul"
         windowtype="Browser:MigrationWizard"
         title="&migrationWizard.title;"
         onload="MigrationWizard.init()"
         onunload="MigrationWizard.uninit()"
         style="width: 40em;"
         buttons="accept,cancel"
         branded="true">
-<wizard>
 
-  <script src="chrome://global/content/customElements.js"/>
-  <script src="chrome://browser/content/migration/migration.js"/>
+<script src="chrome://global/content/customElements.js"/>
+<script src="chrome://browser/content/migration/migration.js"/>
 
+<wizard>
   <wizardpage id="importSource" pageid="importSource" next="selectProfile"
               label="&importSource.title;">
 #ifdef XP_WIN
     <description id="importAll" control="importSourceGroup">&importFrom.label;</description>
 #else
     <description id="importAll" control="importSourceGroup">&importFromUnix.label;</description>
 #endif
     <description id="importBookmarks" control="importSourceGroup" hidden="true">&importFromBookmarks.label;</description>
diff --git a/toolkit/content/widgets/wizard.js b/toolkit/content/widgets/wizard.js
--- a/toolkit/content/widgets/wizard.js
+++ b/toolkit/content/widgets/wizard.js
@@ -12,19 +12,29 @@
   );
   const { Services } = ChromeUtils.import(
     "resource://gre/modules/Services.jsm"
   );
 
   const kDTDs = ["chrome://global/locale/wizard.dtd"];
 
   class MozWizard extends MozXULElement {
-    constructor() {
-      super();
+    static get inheritedAttributes() {
+      return {
+        ".wizard-buttons": "pagestep,firstpage,lastpage",
+      };
+    }
 
+    connectedCallback() {
+      if (this.delayConnectedCallback() || this.hasConnected) {
+        return;
+      }
+      console.log("connectedCallback wizard");
+
+      this.hasConnected = true;
       this._accessMethod = null;
       this._currentPage = null;
       this._canAdvance = true;
       this._canRewind = false;
       this._hasLoaded = false;
       this._pageStack = [];
 
       this._bundle = Services.strings.createBundle(
@@ -41,16 +51,17 @@
             !event.defaultPrevented
           ) {
             this.cancel();
           }
         },
         { mozSystemGroup: true }
       );
 
+
       this.attachShadow({ mode: "open" }).appendChild(
         MozXULElement.parseXULToFragment(`
         <html:link rel="stylesheet" href="chrome://global/content/widgets.css" />
         <hbox class="wizard-header"></hbox>
         <deck class="wizard-page-box" flex="1">
           <slot xmlns="http://www.w3.org/1999/xhtml" name="wizardpage"></slot>
         </deck>
         <slot xmlns="http://www.w3.org/1999/xhtml"></slot>
@@ -81,28 +92,18 @@
            <vbox class="wizard-header-box-text" flex="1">
              <label class="wizard-header-label"/>
              <label class="wizard-header-description"/>
            </vbox>
            <image class="wizard-header-icon"/>
          </hbox>`
         )
       );
-    }
 
-    static get inheritedAttributes() {
-      return {
-        ".wizard-buttons": "pagestep,firstpage,lastpage",
-      };
-    }
-
-    connectedCallback() {
-      if (this.delayConnectedCallback()) {
-        return;
-      }
+      console.log("appended contents into wizard");
 
       document.documentElement.setAttribute("role", "dialog");
 
       this.pageCount = this.wizardPages.length;
 
       this._initPages();
       this.advance(); // start off on the first page
 
@@ -433,19 +434,21 @@
     }
   }
 
   customElements.define("wizard", MozWizard);
 
   class MozWizardPage extends MozXULElement {
     constructor() {
       super();
+      console.log("constructor wizardpage");
       this.pageIndex = -1;
     }
     connectedCallback() {
+      console.log("connected wizardpage");
       this.setAttribute("slot", "wizardpage");
     }
     get pageid() {
       return this.getAttribute("pageid");
     }
     set pageid(val) {
       this.setAttribute("pageid", val);
     }
@@ -457,17 +460,22 @@
       this.parentNode._accessMethod = "random";
       return val;
     }
   }
 
   customElements.define("wizardpage", MozWizardPage);
 
   class MozWizardButtons extends MozXULElement {
+    constructor() {
+      super();
+      console.log("constructor wizardbuttons");
+    }
     connectedCallback() {
+      console.log("connected wizardbuttons");
       this._wizard = this.getRootNode().host;
 
       this.textContent = "";
       this.appendChild(MozXULElement.parseXULToFragment(this._markup, kDTDs));
 
       this._wizardButtonDeck = this.querySelector(".wizard-next-deck");
 
       this.initializeAttributeInheritance();
diff --git a/toolkit/profile/content/createProfileWizard.xul b/toolkit/profile/content/createProfileWizard.xul
--- a/toolkit/profile/content/createProfileWizard.xul
+++ b/toolkit/profile/content/createProfileWizard.xul
@@ -14,26 +14,26 @@
 
 <window id="createProfileWizard"
         title="&newprofile.title;"
         xmlns:html="http://www.w3.org/1999/xhtml"
         xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
         xmlns="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul"
         onload="initWizard();"
         style="&window.size;">
+
+<script src="chrome://global/content/customElements.js"/>
+<script src="chrome://global/content/globalOverlay.js"/>
+<script src="chrome://global/content/editMenuOverlay.js"/>
+<script src="chrome://mozapps/content/profile/createProfileWizard.js"/>
+
 <wizard>
-  <script src="chrome://global/content/customElements.js"/>
-  <script src="chrome://global/content/globalOverlay.js"/>
-  <script src="chrome://global/content/editMenuOverlay.js"/>
-
   <stringbundle id="bundle_profileManager"
                 src="chrome://mozapps/locale/profile/profileSelection.properties"/>
 
-  <script src="chrome://mozapps/content/profile/createProfileWizard.js"/>
-
   <wizardpage id="explanation">
     <description>&profileCreationExplanation_1.text;</description>
     <description>&profileCreationExplanation_2.text;</description>
     <description>&profileCreationExplanation_3.text;</description>
     <spacer flex="1"/>
 #ifdef XP_MACOSX
     <description>&profileCreationExplanation_4Mac.text;</description>
 #else
