diff --git a/browser/components/migration/content/migration.xul b/browser/components/migration/content/migration.xul
--- a/browser/components/migration/content/migration.xul
+++ b/browser/components/migration/content/migration.xul
@@ -16,11 +16,11 @@
         style="width: 40em;"
         buttons="accept,cancel"
         branded="true">
-<wizard>
 
-  <script src="chrome://global/content/customElements.js"/>
-  <script src="chrome://browser/content/migration/migration.js"/>
+<script src="chrome://global/content/customElements.js"/>
+<script src="chrome://browser/content/migration/migration.js"/>
 
+<wizard>
   <wizardpage id="importSource" pageid="importSource" next="selectProfile"
               label="&importSource.title;">
 #ifdef XP_WIN
diff --git a/toolkit/content/widgets/wizard.js b/toolkit/content/widgets/wizard.js
--- a/toolkit/content/widgets/wizard.js
+++ b/toolkit/content/widgets/wizard.js
@@ -16,15 +16,28 @@
 
   const kDTDs = ["chrome://global/locale/wizard.dtd"];
 
+  // Note: MozWizard currently support adding, but not removing MozWizardPage
+  //       children.
   class MozWizard extends MozXULElement {
     constructor() {
       super();
 
+      // About this._accessMethod:
+      //   There are two possible access methods: "sequential" and "random".
+      //   "sequential" causes the MozWizardPage's to be displayed in the order
+      //   that they are added to the DOM.
+      //   The "random" method name is a bit misleading since the pages aren't
+      //   displayed in a random order. Instead, each MozWizardPage must have
+      //   a "next" attribute containing the id of the MozWizardPage that should
+      //   be loaded next.
       this._accessMethod = null;
       this._currentPage = null;
       this._canAdvance = true;
       this._canRewind = false;
       this._hasLoaded = false;
+      this._hasStarted = false; // Whether any MozWizardPage has been shown yet
+      this._wizardButtonsReady = false;
+      this.pageCount = 0;
       this._pageStack = [];
 
       this._bundle = Services.strings.createBundle(
@@ -95,16 +108,8 @@
     }
 
     connectedCallback() {
-      if (this.delayConnectedCallback()) {
-        return;
-      }
-
       document.documentElement.setAttribute("role", "dialog");
-
-      this.pageCount = this.wizardPages.length;
-
-      this._initPages();
-      this.advance(); // start off on the first page
+      this._maybeStartWizard();
 
       window.addEventListener("close", event => {
         if (this.cancel()) {
@@ -176,33 +181,12 @@
       // Setting this attribute allows wizard's clients to dynamically
       // change the styles of each page based on purpose of the page.
       this.setAttribute("currentpageid", val.pageid);
-      if (this.onFirstPage) {
-        this.canRewind = false;
-        this.setAttribute("firstpage", "true");
-        if (AppConstants.platform == "linux") {
-          this.getButton("back").setAttribute("hidden", "true");
-        }
-      } else {
-        this.canRewind = true;
-        this.setAttribute("firstpage", "false");
-        if (AppConstants.platform == "linux") {
-          this.getButton("back").setAttribute("hidden", "false");
-        }
-      }
 
-      if (this.onLastPage) {
-        this.canAdvance = true;
-        this.setAttribute("lastpage", "true");
-      } else {
-        this.setAttribute("lastpage", "false");
-      }
+      this._initCurrentPage();
 
       this._deck.setAttribute("selectedIndex", val.pageIndex);
       this._advanceFocusToPage(val);
 
-      this._adjustWizardHeader();
-      this._wizardButtons.onPageChange();
-
       this._fireEvent(val, "pageshow");
 
       return val;
@@ -357,6 +341,32 @@
       return false;
     }
 
+    _initCurrentPage() {
+      if (this.onFirstPage) {
+        this.canRewind = false;
+        this.setAttribute("firstpage", "true");
+        if (AppConstants.platform == "linux") {
+          this.getButton("back").setAttribute("hidden", "true");
+        }
+      } else {
+        this.canRewind = true;
+        this.setAttribute("firstpage", "false");
+        if (AppConstants.platform == "linux") {
+          this.getButton("back").setAttribute("hidden", "false");
+        }
+      }
+
+      if (this.onLastPage) {
+        this.canAdvance = true;
+        this.setAttribute("lastpage", "true");
+      } else {
+        this.setAttribute("lastpage", "false");
+      }
+
+      this._adjustWizardHeader();
+      this._wizardButtons.onPageChange();
+    }
+
     _advanceFocusToPage(aPage) {
       if (!this._hasLoaded) {
         return;
@@ -375,17 +385,37 @@
       }
     }
 
-    _initPages() {
-      var meth = "sequential";
-      var pages = this.wizardPages;
-      for (var i = 0; i < pages.length; ++i) {
-        var page = pages[i];
-        page.pageIndex = i;
-        if (page.next != "") {
-          meth = "random";
-        }
+    _registerPage(aPage) {
+      aPage.pageIndex = this.pageCount;
+      this.pageCount += 1;
+      if (!this._accessMethod) {
+        this._accessMethod = aPage.next == "" ? "sequential" : "random";
+      }
+      if (!this._maybeStartWizard()) {
+        // If the wizard has already started, adding a page might require
+        // updating elements to reflect that (ex: changing the Finish button to
+        // the Next button).
+        this._initCurrentPage();
       }
-      this._accessMethod = meth;
+    }
+
+    _onWizardButtonsReady() {
+      this._wizardButtonsReady = true;
+      this._maybeStartWizard();
+    }
+
+    _maybeStartWizard(aIsConnected) {
+      if (
+        !this._hasStarted &&
+        this.isConnected &&
+        this._wizardButtonsReady &&
+        this.pageCount > 0
+      ) {
+        this._hasStarted = true;
+        this.advance();
+        return true;
+      }
+      return false;
     }
 
     _adjustWizardHeader() {
@@ -442,6 +472,11 @@
     }
     connectedCallback() {
       this.setAttribute("slot", "wizardpage");
+
+      let wizard = this.closest("wizard");
+      if (wizard) {
+        wizard._registerPage(this);
+      }
     }
     get pageid() {
       return this.getAttribute("pageid");
@@ -486,6 +521,8 @@
           btn.addEventListener("command", listener);
         }
       }
+
+      this._wizard._onWizardButtonsReady();
     }
 
     static get inheritedAttributes() {
diff --git a/toolkit/profile/content/createProfileWizard.xul b/toolkit/profile/content/createProfileWizard.xul
--- a/toolkit/profile/content/createProfileWizard.xul
+++ b/toolkit/profile/content/createProfileWizard.xul
@@ -19,16 +19,16 @@
         xmlns="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul"
         onload="initWizard();"
         style="&window.size;">
+
+<script src="chrome://global/content/customElements.js"/>
+<script src="chrome://global/content/globalOverlay.js"/>
+<script src="chrome://global/content/editMenuOverlay.js"/>
+<script src="chrome://mozapps/content/profile/createProfileWizard.js"/>
+
 <wizard>
-  <script src="chrome://global/content/customElements.js"/>
-  <script src="chrome://global/content/globalOverlay.js"/>
-  <script src="chrome://global/content/editMenuOverlay.js"/>
-
   <stringbundle id="bundle_profileManager"
                 src="chrome://mozapps/locale/profile/profileSelection.properties"/>
 
-  <script src="chrome://mozapps/content/profile/createProfileWizard.js"/>
-
   <wizardpage id="explanation">
     <description>&profileCreationExplanation_1.text;</description>
     <description>&profileCreationExplanation_2.text;</description>

