diff --git a/browser/base/content/tabbrowser.js b/browser/base/content/tabbrowser.js
--- a/browser/base/content/tabbrowser.js
+++ b/browser/base/content/tabbrowser.js
@@ -5252,6 +5252,10 @@
           return;
         }
 
+        if (!browser.docShell || event.target != browser.docShell.document) {
+          return;
+        }
+
         var titleChanged = this.setTabTitle(tab);
         if (titleChanged && !tab.selected && !tab.hasAttribute("busy")) {
           tab.setAttribute("titlechanged", "true");
diff --git a/toolkit/content/browser-child.js b/toolkit/content/browser-child.js
--- a/toolkit/content/browser-child.js
+++ b/toolkit/content/browser-child.js
@@ -29,7 +29,12 @@
 addEventListener(
   "DOMTitleChanged",
   function(aEvent) {
-    if (!aEvent.isTrusted || aEvent.target.defaultView != content) {
+    if (
+      !aEvent.isTrusted ||
+      aEvent.target.defaultView != content ||
+      content.closed ||
+      docShell.document != aEvent.target
+    ) {
       return;
     }
     sendAsyncMessage("DOMTitleChanged", { title: content.document.title });

