# HG changeset patch
# Parent  4f12d77b4f9b6adaf06615c1c8cdc14de836dc1a

diff --git a/devtools/client/themes/webconsole.css b/devtools/client/themes/webconsole.css
--- a/devtools/client/themes/webconsole.css
+++ b/devtools/client/themes/webconsole.css
@@ -1128,8 +1128,42 @@ html[dir="rtl"] .webconsole-output-wrapp
 .theme-twisty {
   cursor: default;
 }
 
 #split-console-close-button::before {
   fill: var(--theme-toolbar-photon-icon-color);
   background-image: var(--close-button-image);
 }
+
+
+.webconsole-output-wrapper {
+  max-height: 100vh !important;
+  height: 100vh !important;
+}
+
+.webconsole-flex-wrapper {
+  display: flex;
+  flex-direction: column;
+  height: 100vh;
+  max-height: 100vh;
+}
+
+.webconsole-filteringbar-wrapper {
+  flex-shrink: 0;
+}
+
+.webconsole-flex-wrapper .webconsole-output {
+  flex: 0 1 auto;
+}
+
+.jsterm-cm .webconsole-flex-wrapper .jsterm-input-container {
+  max-height: unset;
+}
+
+.webconsole-output:not(:empty) {
+  min-height: 19px;
+}
+
+.webconsole-output-wrapper .jsterm-input-container {
+  flex-grow: 1;
+  flex-shrink: 0;
+}
diff --git a/devtools/client/webconsole/components/App.js b/devtools/client/webconsole/components/App.js
--- a/devtools/client/webconsole/components/App.js
+++ b/devtools/client/webconsole/components/App.js
@@ -138,39 +138,41 @@ class App extends Component {
     // * NotificationBox - Notifications for JSTerm (self-xss warning at the moment)
     // * JSTerm - Input command line.
     return (
       div({
         className: classNames.join(" "),
         ref: node => {
           this.node = node;
         }},
-        FilterBar({
-          hidePersistLogsCheckbox: hud.isBrowserConsole,
-          serviceContainer: {
-            attachRefToHud
-          },
-          closeSplitConsole
-        }),
-        ConsoleOutput({
-          serviceContainer,
-          onFirstMeaningfulPaint,
-        }),
-        SideBar({
-          serviceContainer,
-        }),
-        NotificationBox({
-          id: "webconsole-notificationbox",
-          notifications,
-        }),
-        JSTerm({
-          hud,
-          serviceContainer,
-          onPaste: this.onPaste,
-          codeMirrorEnabled: jstermCodeMirror,
+        div({className: "webconsole-flex-wrapper"},
+          FilterBar({
+            hidePersistLogsCheckbox: hud.isBrowserConsole,
+            serviceContainer: {
+              attachRefToHud
+            },
+            closeSplitConsole
+          }),
+          ConsoleOutput({
+            serviceContainer,
+            onFirstMeaningfulPaint,
+          }),
+          NotificationBox({
+            id: "webconsole-notificationbox",
+            notifications,
+          }),
+          JSTerm({
+            hud,
+            serviceContainer,
+            onPaste: this.onPaste,
+            codeMirrorEnabled: jstermCodeMirror,
+          }),
+        ),
+        SideBar({
+          serviceContainer,
         }),
       )
     );
   }
 }
 
 const mapStateToProps = state => ({
   notifications: getAllNotifications(state),
